core/libp2p/basic/message_read_writer.hpp

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_MESSAGE_READ_WRITER_HPP
+#define KAGOME_MESSAGE_READ_WRITER_HPP
+
+#include <memory>
+#include <vector>
+
+#include "common/buffer.hpp"
+#include "libp2p/basic/message_read_writer_error.hpp"
+#include "libp2p/basic/readwriter.hpp"
+#include "libp2p/basic/varint_reader.hpp"
+#include "libp2p/multi/uvarint.hpp"
+
+namespace libp2p::basic {
+  /**
+   * Allows to read and write messages of (\tparam Message) type
+   * @tparam Message - type of the message; for now, works with Protobuf
+   * messages only
+   */
+  template <typename Message>
```

Comment:
Maybe call it ProtoMessageWriter and get rid of template parameter? Anyway you use protobuf specific syntax when you do that
msg.**ParseFromArray**(msg_buf->data(), msg_buf->size());